### ISSUE TITLE
Drop support for pre-Android 4.0 (API 14) versions

### ISSUE THREAD

**Issue by simonpoole** [MarcusWolschon/osmeditor4android#305]
This is a medium term decision and then implementation task.

Currently we provide support for devices going back to Android 2.2 via the ActionBar Sherlock library. ABS was an independent undertaking to provide backwards compatibility mainly for the ActionBar in earlier Android versions. Development of ABS has been stopped and while there is now a partial replacement from google for the ActionBar part of ABS it doesn't cover other aspects of backwards compatibility with older devices that we require. 

Essentially to drop ABS and switch to the google compatibility libraries we would likely have to drop at least pre-Honeycomb devices, but given that Android 3.0 devices essentially don't exist retaining those probably doesn't make sense either.

Numbers: http://opensignal.com/reports/2014/android-fragmentation/

In other words this would currently still imply dropping support for roughly 15% of all devices.

**Comment by johnjohndoe** [comment]
Can you please point out which features are missing from the Toolbar which have been satisfied from ActionBarSherlock in the past?

**Comment by simonpoole** [comment]
ABS provides Fragment support and numerous other things and not just the Actionbar. Toolbar is extremely new and would need a target change(aka API 21 vs 16). I can't remember right now what exactly was the issue when I tested because on paper the recent compatibility library should cover the same, but there were some "roadblocks". Roadblocks always just being a small matter of programming but the implication is spending more time on supporting essentially defunct stuff (iirc SherlockPreferenceActivity was one of many small blockers).

**Comment by niqueco** [comment]
[AppCompat](https://developer.android.com/tools/support-library/features.html#v7-appcompat) (with either [Toolbar](https://developer.android.com/reference/android/support/v7/widget/Toolbar.html) or [AppCompatActivity](https://developer.android.com/reference/android/support/v7/app/AppCompatActivity.html)) doesn't have everything, but with a bit of work can replace ABS. I've started that change locally and it was working. I can finish that if it helps.

It supports API level 7, so no need to drop support for anyone.

**Comment by johnjohndoe** [comment]
@niqueco I could do a review of the migration if you are interested. Please put all commits on a separate branch and make sure each is compiling. Please keep them small.

**Comment by simonpoole** [comment]
@niqueco what would the net value of replacing one library with another be? The whole point of the question is that we currently have extra code and testing complexity due to supporting very early devices. Or to put it in different words: a lot of work. 

Simply changing the library on it own makes things worse: the effort changing the library, plus significant amount of testing due to the change and without doubt bug fixing and workarounds, without reducing the work for continued development. The only net gain would be that we wouldn't have to include ABS as a library project (which is one click of a checkbox).. 

Or to put it differently: the issue is not the/which compatibility library per se, its the fact that we are supporting ancient android versions at all. If the net result of the discussion is that we continue with the support, then I would suggest staying with ABS for the next year or two.

Now if you are looking for things that need refactoring without immediate gains, I can point to a number of things that would make a whole of a lot more sense. For example redoing all dialogs as FragmentDialogs, probably separating out those that are really parts of the core UI from those that are really just alerts.

**Comment by johnjohndoe** [comment]
Agreed. If there is no benefit it does not make sense.
Regarding the dialogs please link the relevant issues in your message.

**Comment by MarcusWolschon** [comment]
I could live with supporting only API7+ but any change of this magnitude should have some actual gain for the user. I can't see that in this proposal. Dropping support for Android 3 tablets and Android 2.x phones is certainly not a feature. Many old phone are still used for tasks like mapping. You can hand them out and not worry about expensive, modern devices disapearing

**Comment by simonpoole** [comment]
Sorry for not responding earlier. 

I've started compiling a list of tasks that have low or actually no dependency on other stuff in Vespucci and are reasonably self-contained: https://github.com/MarcusWolschon/osmeditor4android/labels/Task with a bit more detail in the description to give some guideance. 

If somebody starts work on one I would change the Assignee to them (just as with any other open issue).

**Comment by johnjohndoe** [comment]
@simonpoole Thank you for the task descriptions.

**Comment by simonpoole** [comment]
@MarcusWolschon that is naturally why I wanted the discussion, so that we can reaffirm the commitment to further support or drop it in a structured long term way (which would likely not be before 2017 in any case),

**Comment by johnjohndoe** [comment]
Please note that you will [run into errors](http://stackoverflow.com/questions/31769937/attribute-titletextstyle-has-already-been-defined) when you add `appcompat-v7` or libraries which depend on it while `actionbarsherlock` is still there.

**Comment by simonpoole** [comment]
I tried updating to the current v4 support library (23.1) this however seems to break ABS definitely. It seems as if this might be essentially google forcing us to change .@niqueco any chance of a PR of the work you did vs. 0.9.7?

**Comment by simonpoole** [comment]
It seems as if a large number of changes were made to the compat lib in 22, so it is unlikely that it makes a lot of sense to try to use old patches. In any case I'm opening a new issue for the specific task: https://github.com/MarcusWolschon/osmeditor4android/issues/354

**Comment by niqueco** [comment]
Yes, I could work on upgrading to the latest AppCompat. It shouldn't take long.

**Comment by johnjohndoe** [comment]
@niqueco :+1:

**Comment by simonpoole** [comment]
@niqueco I've added some comments to https://github.com/MarcusWolschon/osmeditor4android/issues/354 

It should be noted for this issue that switching to AppCompat puts us completely at the mercy of google for backwards compatibility and no longer makes it only our decision.

**Comment by simonpoole** [comment]
FYI the appcompat branch contains a version using appcompat instead of ABS, see  https://github.com/MarcusWolschon/osmeditor4android/issues/354 This version uses support library v23 (Marshmallow) so hopefully will not need all too much work going forward, depends naturally on how much stuff google breaks. 

I would expect that this will be merged in a week or two, if no further large roadblocks turn up.

**Comment by simonpoole** [comment]
I merged the version with appcompat a while back, due to the changes in https://github.com/MarcusWolschon/osmeditor4android/commit/b1e5c9372abf5d3e581511371b36fc30e1e46892 ACRA is causing issues with API7 - Android 2.1, given that we haven't advertize support for anything below 2.2 for a long time, I don't believe this is an issue.

**Comment by simonpoole** [comment]
The time is running out on Android 2.2 / API 8
I'm working on a branch that when merged will compile against API 24 (N). Support libraries version 24.1.1 are the last that still support 2.2 (from roughly mid last year), the last 24 version was 24.2.1 which we can't use without loosing 2.2 support, not to mention the current 25 libs.

**Comment by MarcusWolschon** [comment]
support Libs 24.2.0 had heavy bugs with VectorDrawables and with ErrorIcons in TextViews.

24.2.1-25.0.0 had a crashing bug "LinearLayout$LayoutParams cannot be cast to CoordinatorLayout$LayoutParams".
I don't know if that's fixed in later 25.x.x versions yet. It COULD be fixed since 25.1.0 .

These things are something to look out for.

**Comment by simonpoole** [comment]
The support library 26 (Android 8) drops support for Android versions before 4.0 I don't know of any immediate version to change at this point in time, but it is something to be aware of.

**Comment by simonpoole** [comment]
Seems as if we should actually be dropping support for everything before 4.1, see https://github.com/MarcusWolschon/osmeditor4android/issues/672

Currently this is likely amount to slightly below 2% of all android devices

**Comment by johnjohndoe** [comment]
@simonpoole Also mind the [changes with regards to Google Play](https://commonsware.com/blog/2018/01/08/android-version-ratchet.html) by the end of 2018.

**Comment by simonpoole** [comment]
@johnjohndoe yes that was clear when the goog announced the new policy, the trade rags (heise etc) got it wrong as normal which conveniently for the goog makes it all the devs fault in the eye of the public.

**Comment by johnjohndoe** [comment]
@simonpoole Can you please reference the post where the announcement was misunderstood?

**Comment by simonpoole** [comment]
https://github.com/MarcusWolschon/osmeditor4android
